### ISSUE TITLE
Resampling for oneclass-classification/anomaly detection

### ISSUE THREAD

**Issue by MinhAnhL** [mlr-org/mlr#1852]
In oneclass-classification you usually train on data, which only contain one class (the normal class). The trained model should learn the behaviour of the normal class and therefore it should be able to detect anomalies in the test data (if there are any).
To evaluate the performance we would use resampling methods, but mlr does not provide the option yet to only have oneclass in the resampled training data.

Suggestions on how to implement resampling for oneclass:

**1. Hold out** 
- split the data, with only normal data in the train set
- and all anomalies in the test set

**2. k-CV**
- split the data as usual into k-partition
- one partition is the test set as usual
- the rest k-1 partition are the train set as usual, BUT drop anomalies when training the model
- maybe per default stratify == TRUE ? (to make sure we have anomalies in the test data)

**3. Bootstrapping**
- sample with replacement only from the normal data to create train set
- anomaly data are always in the test set.

**4. Subsampling** 
- nearly analog to Bootstrapping
- sample without replacement only from the normal data to create train set (size is $split = a proportion of training cases e.g. 2/3)
- anomaly data are always in the test set.
- MAYBE:  for test set: sample without replacement of anomaly data with size $split (e.g. 2/3) to get some variation in the anomaly data and also to keep the proportion between number of anomalies and number of normal data?

**5. repeat CV**
- stays the same as instantiateResampleInstance.RepCVDesc() in mlr but with one line changed
instead of d = makeResampleDesc("CV", iters = folds) use method 2 from above ("OC-CV")

**6. LOO** 
- not necessary(doesn't make sense to make an own class for oneclass 

**Add-on question:** Is it necessary to have a resampling option that creates test data without anomalies?
I particularly think of the question: Does the Anomaly Detection model learned the normal class well enough to recognize all normal observation as normal in a dataset which only contain normal classes. Or does the model always detect observations with the lowest score as anomalies, although there aren't any/ Test false alarm.
Alternative: Only input normal data, therefore test and train only on normal observation
But, what if: train on data which has normal and anomaly cases but test only on normal data.? Not sure whether this is a common case and worth analysing.

**Comment by mb706** [comment]
Things that I can think of:

> MAYBE: for test set: sample without replacement of anomaly data with size $split (e.g. 2/3)

Depending on the measure, if you always put all the anomalies in the test set there are going to be bias problems with most of these.

> resampling option that creates test data without anomalies

Shouldn't the prediction of one row, in general, be independent of all the other rows to predict? I.e. should `rbind(predict(model, newdata = data[subset, ])$data, predict(model, newdata = data[-subset, ])$data)` be the same as `predict(model, newdata = data)$data`? If that is the case, you probably just want a different kind of measure instead of a different resampling method (?). Or do some of the one-class learners do online learning / prediction dependent on prediction input?

> train on data which has normal and anomaly cases

That sounds like two-class learning to me.

**Comment by berndbischl** [comment]
> Depending on the measure, if you always put all the anomalies in the test set there are going to be bias problems with most of these.

well. for anomaly detection only ROC like measures make sense anyway, right?
so i would explain this in the docs.
proceed as outlined by @MinhAnhL, as i think using the scarce anom data as much as possible is best.

if anybody knows any papers about the general resampling topic, please post! 
we dont find anything

**Comment by berndbischl** [comment]
> train on data which has normal and anomaly cases
> That sounds like two-class learning to me.

its not 2class. the standarrd example is:
10K obs class "normal", 5obs in class "anom".
and assume that i have the labels.

now that is not truely binary classification, because i cannot really train on this with the normal algos?
so i normally do one-class with the 10K obs but somehow try to adapt the algo a bit, so it can use the 5 anom obs, to make the model better.

**Comment by berndbischl** [comment]
> resampling option that creates test data without anomalies
> Shouldn't the prediction of one row, in general, be independent of all the other rows to predict? I.e. should rbind(predict(model, newdata = data[subset, ])$data, predict(model, newdata = data[-subset, ])$data) be the same as predict(model, newdata = data)$data? If that is the case, you probably just want a different kind of measure instead of a different resampling method (?). Or do some of the one-class learners do online learning / prediction dependent on prediction input?

i guess martin is correct here. one could simply ignores these unwanted predictions, e.g. in a measure
